File: exercises/c/01_lexical_analyzer/lexical_analyzer.c
/*
================================================================================
Exercise 1: Lexical Analyzer (Scanner)  --  C Implementation
================================================================================

CONCEPT:
    A lexical analyzer (also called a scanner or tokenizer) is the first phase
    of a compiler. It reads the source program character by character and groups
    them into meaningful sequences called "lexemes." Each lexeme is mapped to a
    token -- a (token-type, attribute-value) pair that is passed to the parser.

ALGORITHM:
    The scanner iterates through the source string one character at a time.
    Based on the current character it decides which class of token to build:
      - Letters/underscore  -> identifier or keyword
      - Digits              -> integer or float constant
      - Quote (")           -> string literal
      - Operator characters -> single or double operator
      - Punctuation         -> delimiter
      - '/' followed by '/' -> comment (skipped)
    A lookahead of one character is used for two-character operators (==, !=,
    etc.) and for detecting comments.  After each lexeme is extracted it is
    classified: if it matches a reserved word it becomes KEYWORD, otherwise
    IDENTIFIER, and so on.

COMPLEXITY:
    Time:  O(n) -- each character visited at most twice (read + lookahead).
    Space: O(n) -- token list storage.

EXAMPLE:
    Input:  "int x = 10;"
    Output: KEYWORD(int)  IDENTIFIER(x)  OPERATOR(=)  INTEGER(10)  PUNCT(;)
================================================================================
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/* ---- constants ---------------------------------------------------------- */
#define MAX_TOKENS     1024
#define MAX_LEXEME_LEN 256
#define MAX_SOURCE_LEN 4096

/* ---- token types -------------------------------------------------------- */
typedef enum {
    TOKEN_KEYWORD,
    TOKEN_IDENTIFIER,
    TOKEN_INTEGER,
    TOKEN_FLOAT,
    TOKEN_STRING,
    TOKEN_OPERATOR,
    TOKEN_PUNCTUATION,
    TOKEN_UNKNOWN
} TokenType;

static const char *token_type_names[] = {
    "KEYWORD",
    "IDENTIFIER",
    "INTEGER_CONST",
    "FLOAT_CONST",
    "STRING_LITERAL",
    "OPERATOR",
    "PUNCTUATION",
    "UNKNOWN"
};

/* ---- token structure ---------------------------------------------------- */
typedef struct {
    TokenType type;
    char      lexeme[MAX_LEXEME_LEN];
    int       line;
    int       col;
} Token;

/* ---- keyword table ------------------------------------------------------ */
static const char *keywords[] = {
    "int", "float", "char", "double", "void",
    "if", "else", "while", "for", "do",
    "return", "break", "continue",
    "struct", "typedef", "enum", "switch", "case", "default",
    "main", "printf", "scanf", "include", "define",
    NULL  /* sentinel */
};

/* ---- helper: is this string a keyword? ---------------------------------- */
static int is_keyword(const char *word)
{
    for (int i = 0; keywords[i] != NULL; i++) {
        if (strcmp(word, keywords[i]) == 0)
            return 1;
    }
    return 0;
}

/* ---- helper: is character a single-char operator? ----------------------- */
static int is_operator_char(char ch)
{
    return ch == '+' || ch == '-' || ch == '*' || ch == '/' ||
           ch == '%' || ch == '=' || ch == '<' || ch == '>' ||
           ch == '!' || ch == '&' || ch == '|' || ch == '^' || ch == '~';
}

/* ---- helper: is character punctuation? ---------------------------------- */
static int is_punctuation(char ch)
{
    return ch == ';' || ch == ',' || ch == '(' || ch == ')' ||
           ch == '{' || ch == '}' || ch == '[' || ch == ']' ||
           ch == '#' || ch == ':' || ch == '.';
}

/* ---- add a token to the list -------------------------------------------- */
static int add_token(Token tokens[], int *count, TokenType type,
                     const char *lexeme, int line, int col)
{
    if (*count >= MAX_TOKENS) return -1;
    tokens[*count].type = type;
    strncpy(tokens[*count].lexeme, lexeme, MAX_LEXEME_LEN - 1);
    tokens[*count].lexeme[MAX_LEXEME_LEN - 1] = '\0';
    tokens[*count].line = line;
    tokens[*count].col  = col;
    (*count)++;
    return 0;
}

/* ======================================================================== */
/*  tokenize -- the main scanning function                                  */
/* ======================================================================== */
static int tokenize(const char *src, Token tokens[])
{
    int count  = 0;
    int pos    = 0;
    int length = (int)strlen(src);
    int line   = 1;
    int col    = 1;

    while (pos < length) {
        char ch = src[pos];

        /* ---- whitespace ------------------------------------------------ */
        if (ch == ' ' || ch == '\t' || ch == '\r') {
            pos++; col++;
            continue;
        }
        if (ch == '\n') {
            pos++; line++; col = 1;
            continue;
        }

        /* ---- single-line comment  //  ---------------------------------- */
        if (ch == '/' && pos + 1 < length && src[pos + 1] == '/') {
            while (pos < length && src[pos] != '\n') pos++;
            continue;
        }

        /* ---- multi-line comment  / * ... * /  -------------------------- */
        if (ch == '/' && pos + 1 < length && src[pos + 1] == '*') {
            pos += 2; col += 2;
            while (pos + 1 < length) {
                if (src[pos] == '*' && src[pos + 1] == '/') { pos += 2; col += 2; break; }
                if (src[pos] == '\n') { line++; col = 1; } else { col++; }
                pos++;
            }
            continue;
        }

        /* ---- string literal  "..."  ------------------------------------ */
        if (ch == '"') {
            int start_col = col;
            char buf[MAX_LEXEME_LEN];
            int bi = 0;
            buf[bi++] = ch; pos++; col++;
            while (pos < length && src[pos] != '"' && bi < MAX_LEXEME_LEN - 2) {
                if (src[pos] == '\\') { buf[bi++] = src[pos]; pos++; col++; }
                if (pos < length) { buf[bi++] = src[pos]; pos++; col++; }
            }
            if (pos < length) { buf[bi++] = src[pos]; pos++; col++; } /* closing " */
            buf[bi] = '\0';
            add_token(tokens, &count, TOKEN_STRING, buf, line, start_col);
            continue;
        }

        /* ---- character literal  '.'  ----------------------------------- */
        if (ch == '\'') {
            int start_col = col;
            char buf[MAX_LEXEME_LEN];
            int bi = 0;
            buf[bi++] = ch; pos++; col++;
            while (pos < length && src[pos] != '\'' && bi < MAX_LEXEME_LEN - 2) {
                if (src[pos] == '\\') { buf[bi++] = src[pos]; pos++; col++; }
                if (pos < length) { buf[bi++] = src[pos]; pos++; col++; }
            }
            if (pos < length) { buf[bi++] = src[pos]; pos++; col++; }
            buf[bi] = '\0';
            add_token(tokens, &count, TOKEN_STRING, buf, line, start_col);
            continue;
        }

        /* ---- identifiers / keywords ------------------------------------ */
        if (isalpha((unsigned char)ch) || ch == '_') {
            int start_col = col;
            char buf[MAX_LEXEME_LEN];
            int bi = 0;
            while (pos < length &&
                   (isalnum((unsigned char)src[pos]) || src[pos] == '_') &&
                   bi < MAX_LEXEME_LEN - 1) {
                buf[bi++] = src[pos]; pos++; col++;
            }
            buf[bi] = '\0';
            TokenType tt = is_keyword(buf) ? TOKEN_KEYWORD : TOKEN_IDENTIFIER;
            add_token(tokens, &count, tt, buf, line, start_col);
            continue;
        }

        /* ---- numeric constants ----------------------------------------- */
        if (isdigit((unsigned char)ch)) {
            int start_col = col;
            int is_float  = 0;
            char buf[MAX_LEXEME_LEN];
            int bi = 0;
            while (pos < length && isdigit((unsigned char)src[pos]) && bi < MAX_LEXEME_LEN - 1) {
                buf[bi++] = src[pos]; pos++; col++;
            }
            if (pos < length && src[pos] == '.') {
                is_float = 1;
                buf[bi++] = src[pos]; pos++; col++;
                while (pos < length && isdigit((unsigned char)src[pos]) && bi < MAX_LEXEME_LEN - 1) {
                    buf[bi++] = src[pos]; pos++; col++;
                }
            }
            buf[bi] = '\0';
            add_token(tokens, &count, is_float ? TOKEN_FLOAT : TOKEN_INTEGER,
                      buf, line, start_col);
            continue;
        }

        /* ---- two-character operators ----------------------------------- */
        if (pos + 1 < length) {
            char two[3] = { src[pos], src[pos + 1], '\0' };
            /* list of all two-char operators */
            static const char *twos[] = {
                "==", "!=", "<=", ">=", "++", "--",
                "+=", "-=", "*=", "/=", "%=",
                "&&", "||", "<<", ">>", NULL
            };
            int found = 0;
            for (int i = 0; twos[i] != NULL; i++) {
                if (two[0] == twos[i][0] && two[1] == twos[i][1]) {
                    add_token(tokens, &count, TOKEN_OPERATOR, two, line, col);
                    pos += 2; col += 2;
                    found = 1;
                    break;
                }
            }
            if (found) continue;
        }

        /* ---- single-character operators -------------------------------- */
        if (is_operator_char(ch)) {
            char buf[2] = { ch, '\0' };
            add_token(tokens, &count, TOKEN_OPERATOR, buf, line, col);
            pos++; col++;
            continue;
        }

        /* ---- punctuation / delimiters ---------------------------------- */
        if (is_punctuation(ch)) {
            char buf[2] = { ch, '\0' };
            add_token(tokens, &count, TOKEN_PUNCTUATION, buf, line, col);
            pos++; col++;
            continue;
        }

        /* ---- unknown --------------------------------------------------- */
        {
            char buf[2] = { ch, '\0' };
            add_token(tokens, &count, TOKEN_UNKNOWN, buf, line, col);
            pos++; col++;
        }
    }

    return count;
}

/* ---- pretty print ------------------------------------------------------- */
static void print_tokens(const Token tokens[], int count)
{
    /* find max lexeme width */
    int max_lex = 6; /* minimum "LEXEME" header */
    for (int i = 0; i < count; i++) {
        int len = (int)strlen(tokens[i].lexeme);
        if (len > max_lex) max_lex = len;
    }

    printf("+---+-------------------+");
    for (int i = 0; i < max_lex + 2; i++) putchar('-');
    printf("+----------+\n");

    printf("| # | %-17s | %-*s | LOCATION |\n", "TOKEN TYPE", max_lex, "LEXEME");

    printf("+---+-------------------+");
    for (int i = 0; i < max_lex + 2; i++) putchar('-');
    printf("+----------+\n");

    for (int i = 0; i < count; i++) {
        printf("|%3d| %-17s | %-*s | L%d:C%-4d |\n",
               i + 1,
               token_type_names[tokens[i].type],
               max_lex,
               tokens[i].lexeme,
               tokens[i].line,
               tokens[i].col);
    }

    printf("+---+-------------------+");
    for (int i = 0; i < max_lex + 2; i++) putchar('-');
    printf("+----------+\n");
}

static void print_summary(const Token tokens[], int count)
{
    int cats[8] = {0};
    for (int i = 0; i < count; i++)
        cats[tokens[i].type]++;

    printf("\n+-------------------+-------+\n");
    printf("| TOKEN TYPE        | COUNT |\n");
    printf("+-------------------+-------+\n");
    for (int i = 0; i < 8; i++) {
        if (cats[i] > 0)
            printf("| %-17s | %5d |\n", token_type_names[i], cats[i]);
    }
    printf("+-------------------+-------+\n");
    printf("| %-17s | %5d |\n", "TOTAL", count);
    printf("+-------------------+-------+\n");
}

/* ---- default sample code ------------------------------------------------ */
static const char *SAMPLE_CODE =
    "// A small C program for lexical analysis demonstration\n"
    "#include <stdio.h>\n"
    "\n"
    "int main() {\n"
    "    int count = 0;\n"
    "    float ratio = 3.14;\n"
    "    char name[] = \"hello\";\n"
    "\n"
    "    for (int i = 0; i <= 10; i++) {\n"
    "        if (i != 5 && i >= 2) {\n"
    "            count += i;\n"
    "        } else {\n"
    "            count--;\n"
    "        }\n"
    "    }\n"
    "\n"
    "    while (count > 0) {\n"
    "        ratio *= 0.5;\n"
    "        count = count - 1;\n"
    "    }\n"
    "\n"
    "    printf(\"Result: %d\\n\", count);\n"
    "    return 0;\n"
    "}\n";

/* ---- main --------------------------------------------------------------- */
int main(void)
{
    printf("======================================================================\n");
    printf("         LEXICAL ANALYZER  --  Compiler Design Exercise 1\n");
    printf("======================================================================\n\n");

    printf("Choose input mode:\n");
    printf("  1) Use default sample C code\n");
    printf("  2) Enter your own code (end with an empty line)\n");
    printf("\nEnter choice [1]: ");

    char choice_buf[16];
    if (fgets(choice_buf, sizeof(choice_buf), stdin) == NULL)
        choice_buf[0] = '1';

    char source[MAX_SOURCE_LEN];

    if (choice_buf[0] == '2') {
        printf("\nEnter source code (press Enter on an empty line to finish):\n\n");
        source[0] = '\0';
        char line_buf[512];
        while (fgets(line_buf, sizeof(line_buf), stdin) != NULL) {
            if (line_buf[0] == '\n') break;
            strncat(source, line_buf, MAX_SOURCE_LEN - strlen(source) - 1);
        }
    } else {
        strncpy(source, SAMPLE_CODE, MAX_SOURCE_LEN - 1);
        source[MAX_SOURCE_LEN - 1] = '\0';
    }

    /* display source */
    printf("\n--- Source Code ---\n");
    int line_no = 1;
    const char *p = source;
    printf("  %3d | ", line_no);
    while (*p) {
        putchar(*p);
        if (*p == '\n' && *(p + 1)) {
            line_no++;
            printf("  %3d | ", line_no);
        }
        p++;
    }
    printf("\n--- End of Source ---\n\n");

    /* tokenize and display */
    Token tokens[MAX_TOKENS];
    int count = tokenize(source, tokens);

    printf("--- Token Table ---\n");
    print_tokens(tokens, count);
    print_summary(tokens, count);

    return 0;
}

File: exercises/python/01_lexical_analyzer/lexical_analyzer.py
"""
================================================================================
Exercise 1: Lexical Analyzer (Scanner)
================================================================================

CONCEPT:
    A lexical analyzer (also called a scanner or tokenizer) is the first phase
    of a compiler. It reads the source program character by character and groups
    them into meaningful sequences called "lexemes." Each lexeme is mapped to a
    token, which is a <token-type, attribute-value> pair that is passed to the
    subsequent phase (the parser/syntax analyzer).

ALGORITHM:
    The scanner uses a character-by-character approach with a state-driven loop.
    At each step, the current character determines which kind of token is being
    formed. For example, if the character is a letter, the scanner continues
    reading letters and digits to form an identifier or keyword. If it is a
    digit, it reads a numeric constant (integer or float). Operators that can
    be one or two characters (like '=' vs '==') require a one-character lookahead
    to decide the correct token. Comments starting with '//' are detected and
    skipped. String literals are read from one quote to the matching quote.

    After a lexeme is fully read, it is classified:
      - If it matches a reserved keyword, the token type is KEYWORD.
      - Otherwise an alphabetic lexeme is an IDENTIFIER.
      - Numeric lexemes are INTEGER_CONST or FLOAT_CONST.
      - Operator and punctuation lexemes have their own categories.

COMPLEXITY:
    Time:  O(n) where n is the length of the source code -- each character is
           visited at most twice (once for reading, once for lookahead).
    Space: O(n) for storing the resulting list of tokens.

EXAMPLE INPUT:
    int main() {
        int x = 10;
        if (x >= 5) {
            x = x + 1;
        }
        return 0;
    }

EXAMPLE OUTPUT (partial):
    TOKEN TYPE        | LEXEME
    ------------------+---------
    KEYWORD           | int
    KEYWORD           | main
    PUNCTUATION       | (
    PUNCTUATION       | )
    PUNCTUATION       | {
    KEYWORD           | int
    IDENTIFIER        | x
    OPERATOR          | =
    INTEGER_CONST     | 10
    ...
================================================================================
"""

# ---------------------------------------------------------------------------
# Token categories
# ---------------------------------------------------------------------------
TOKEN_KEYWORD       = "KEYWORD"
TOKEN_IDENTIFIER    = "IDENTIFIER"
TOKEN_INTEGER       = "INTEGER_CONST"
TOKEN_FLOAT         = "FLOAT_CONST"
TOKEN_STRING        = "STRING_LITERAL"
TOKEN_OPERATOR      = "OPERATOR"
TOKEN_PUNCTUATION   = "PUNCTUATION"
TOKEN_UNKNOWN       = "UNKNOWN"

# ---------------------------------------------------------------------------
# Language tables
# ---------------------------------------------------------------------------
KEYWORDS = {
    "int", "float", "char", "double", "void",
    "if", "else", "while", "for", "do",
    "return", "break", "continue",
    "struct", "typedef", "enum", "switch", "case", "default",
    "main", "printf", "scanf", "include", "define",
}

# Two-character operators must be checked before single-character ones.
TWO_CHAR_OPERATORS = {
    "==", "!=", "<=", ">=", "++", "--",
    "+=", "-=", "*=", "/=", "%=",
    "&&", "||", "<<", ">>",
}

SINGLE_CHAR_OPERATORS = set("+-*/%=<>!&|^~")

PUNCTUATION = set(";,(){}[]:#.")


# ---------------------------------------------------------------------------
# Lexical Analyzer
# ---------------------------------------------------------------------------
class Token:
    """Represents a single token produced by the lexer."""
    def __init__(self, token_type: str, lexeme: str, line: int, col: int):
        self.token_type = token_type
        self.lexeme = lexeme
        self.line = line
        self.col = col

    def __repr__(self):
        return f"Token({self.token_type}, {self.lexeme!r}, line={self.line})"


def tokenize(source_code: str) -> list:
    """
    Scan *source_code* character by character and return a list of Token objects.
    """
    tokens = []
    pos = 0                # current index in source_code
    length = len(source_code)
    line = 1               # current source line  (for error messages)
    col = 1                # current column

    while pos < length:
        ch = source_code[pos]

        # ---- whitespace ------------------------------------------------
        if ch in " \t\r":
            pos += 1
            col += 1
            continue

        if ch == "\n":
            pos += 1
            line += 1
            col = 1
            continue

        # ---- single-line comment  //  ----------------------------------
        if ch == "/" and pos + 1 < length and source_code[pos + 1] == "/":
            # skip until end of line
            while pos < length and source_code[pos] != "\n":
                pos += 1
            continue  # the newline itself will be handled on next iteration

        # ---- multi-line comment  /* ... */  ----------------------------
        if ch == "/" and pos + 1 < length and source_code[pos + 1] == "*":
            pos += 2
            col += 2
            while pos + 1 < length:
                if source_code[pos] == "*" and source_code[pos + 1] == "/":
                    pos += 2
                    col += 2
                    break
                if source_code[pos] == "\n":
                    line += 1
                    col = 1
                else:
                    col += 1
                pos += 1
            continue

        # ---- string literal  "..."  ------------------------------------
        if ch == '"':
            start_col = col
            start_pos = pos
            pos += 1
            col += 1
            while pos < length and source_code[pos] != '"':
                if source_code[pos] == "\\":  # handle escape sequences
                    pos += 1
                    col += 1
                if source_code[pos] == "\n":
                    line += 1
                    col = 1
                else:
                    col += 1
                pos += 1
            if pos < length:  # skip closing quote
                pos += 1
                col += 1
            lexeme = source_code[start_pos:pos]
            tokens.append(Token(TOKEN_STRING, lexeme, line, start_col))
            continue

        # ---- character literal  '.'  -----------------------------------
        if ch == "'":
            start_col = col
            start_pos = pos
            pos += 1
            col += 1
            while pos < length and source_code[pos] != "'":
                if source_code[pos] == "\\":
                    pos += 1
                    col += 1
                col += 1
                pos += 1
            if pos < length:
                pos += 1
                col += 1
            lexeme = source_code[start_pos:pos]
            tokens.append(Token(TOKEN_STRING, lexeme, line, start_col))
            continue

        # ---- identifiers / keywords ------------------------------------
        if ch.isalpha() or ch == "_":
            start_col = col
            start_pos = pos
            while pos < length and (source_code[pos].isalnum() or source_code[pos] == "_"):
                pos += 1
                col += 1
            lexeme = source_code[start_pos:pos]
            tok_type = TOKEN_KEYWORD if lexeme in KEYWORDS else TOKEN_IDENTIFIER
            tokens.append(Token(tok_type, lexeme, line, start_col))
            continue

        # ---- numeric constants (integer and float) ---------------------
        if ch.isdigit():
            start_col = col
            start_pos = pos
            is_float = False
            while pos < length and source_code[pos].isdigit():
                pos += 1
                col += 1
            # check for decimal point
            if pos < length and source_code[pos] == ".":
                is_float = True
                pos += 1
                col += 1
                while pos < length and source_code[pos].isdigit():
                    pos += 1
                    col += 1
            lexeme = source_code[start_pos:pos]
            tok_type = TOKEN_FLOAT if is_float else TOKEN_INTEGER
            tokens.append(Token(tok_type, lexeme, line, start_col))
            continue

        # ---- preprocessor directive  # --------------------------------
        if ch == "#":
            start_col = col
            tokens.append(Token(TOKEN_PUNCTUATION, "#", line, start_col))
            pos += 1
            col += 1
            continue

        # ---- two-character operators -----------------------------------
        if pos + 1 < length:
            two = source_code[pos:pos + 2]
            if two in TWO_CHAR_OPERATORS:
                tokens.append(Token(TOKEN_OPERATOR, two, line, col))
                pos += 2
                col += 2
                continue

        # ---- single-character operators --------------------------------
        if ch in SINGLE_CHAR_OPERATORS:
            tokens.append(Token(TOKEN_OPERATOR, ch, line, col))
            pos += 1
            col += 1
            continue

        # ---- punctuation / delimiters ----------------------------------
        if ch in PUNCTUATION:
            tokens.append(Token(TOKEN_PUNCTUATION, ch, line, col))
            pos += 1
            col += 1
            continue

        # ---- unknown character -----------------------------------------
        tokens.append(Token(TOKEN_UNKNOWN, ch, line, col))
        pos += 1
        col += 1

    return tokens


# ---------------------------------------------------------------------------
# Pretty-printing
# ---------------------------------------------------------------------------
def print_tokens(tokens: list) -> None:
    """Print tokens in a formatted table."""
    header_type = "TOKEN TYPE"
    header_lex  = "LEXEME"
    header_loc  = "LOCATION"

    # determine column widths
    type_width = max(len(header_type), max((len(t.token_type) for t in tokens), default=0))
    lex_width  = max(len(header_lex),  max((len(t.lexeme) for t in tokens), default=0))
    loc_width  = max(len(header_loc), 12)

    sep = "+" + "-" * (type_width + 2) + "+" + "-" * (lex_width + 2) + "+" + "-" * (loc_width + 2) + "+"

    print(sep)
    print(f"| {header_type:<{type_width}} | {header_lex:<{lex_width}} | {header_loc:<{loc_width}} |")
    print(sep)
    for tok in tokens:
        loc = f"L{tok.line}:C{tok.col}"
        print(f"| {tok.token_type:<{type_width}} | {tok.lexeme:<{lex_width}} | {loc:<{loc_width}} |")
    print(sep)


def print_summary(tokens: list) -> None:
    """Print a summary count of each token category."""
    counts: dict[str, int] = {}
    for tok in tokens:
        counts[tok.token_type] = counts.get(tok.token_type, 0) + 1

    print("\n+---------------------+-------+")
    print(f"| {'TOKEN TYPE':<19} | {'COUNT':>5} |")
    print("+---------------------+-------+")
    for cat in sorted(counts):
        print(f"| {cat:<19} | {counts[cat]:>5} |")
    print("+---------------------+-------+")
    print(f"| {'TOTAL':<19} | {len(tokens):>5} |")
    print("+---------------------+-------+")


# ---------------------------------------------------------------------------
# Default sample input
# ---------------------------------------------------------------------------
SAMPLE_CODE = r"""
// A small C program for lexical analysis demonstration
#include <stdio.h>

int main() {
    int count = 0;
    float ratio = 3.14;
    char name[] = "hello";

    // Loop with multiple operators
    for (int i = 0; i <= 10; i++) {
        if (i != 5 && i >= 2) {
            count += i;
        } else {
            count--;
        }
    }

    while (count > 0) {
        ratio *= 0.5;
        count = count - 1;
    }

    printf("Result: %d\n", count);
    return 0;
}
"""


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
def main():
    print("=" * 70)
    print("         LEXICAL ANALYZER  --  Compiler Design Exercise 1")
    print("=" * 70)

    print("\nChoose input mode:")
    print("  1) Use default sample C code")
    print("  2) Enter your own code (end with a blank line)")

    choice = input("\nEnter choice [1]: ").strip()

    if choice == "2":
        print("\nEnter source code (press Enter twice to finish):\n")
        lines = []
        while True:
            line = input()
            if line == "":
                break
            lines.append(line)
        source = "\n".join(lines)
    else:
        source = SAMPLE_CODE

    print("\n--- Source Code ---")
    for num, line in enumerate(source.split("\n"), start=1):
        print(f"  {num:3d} | {line}")
    print("--- End of Source ---\n")

    tokens = tokenize(source)

    print("--- Token Table ---")
    print_tokens(tokens)
    print_summary(tokens)


if __name__ == "__main__":
    main()
